src/client/cli/cmd/snapshot.cpp

```diff
@@ -0,0 +1,106 @@
+/*
+ * Copyright (C) Canonical, Ltd.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; version 3.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+#include "snapshot.h"
+
+#include "animated_spinner.h"
+#include "common_callbacks.h"
+#include "common_cli.h"
+
+#include <multipass/cli/argparser.h>
+
+namespace mp = multipass;
+namespace cmd = mp::cmd;
+
+mp::ReturnCode cmd::Snapshot::run(mp::ArgParser* parser)
+{
+    if (auto ret = parse_args(parser); ret != ParseCode::Ok)
+        return parser->returnCodeFrom(ret);
+
+    AnimatedSpinner spinner{cout};
+
+    auto on_success = [this, &spinner](mp::SnapshotReply& reply) {
+        spinner.stop();
+        fmt::print(cout, "Snapshot taken: {}.{}\n", request.instance(), reply.snapshot());
+        return ReturnCode::Ok;
+    };
+
+    auto on_failure = [this, &spinner](grpc::Status& status) {
+        spinner.stop();
+        return standard_failure_handler_for(name(), cerr, status);
+    };
+
+    spinner.start("Taking snapshot ");
+    return dispatch(&RpcMethod::snapshot,
+                    request,
+                    on_success,
+                    on_failure,
+                    make_logging_spinner_callback<SnapshotRequest, SnapshotReply>(spinner, cerr));
+}
+
+std::string cmd::Snapshot::name() const
+{
+    return "snapshot";
+}
+
+QString cmd::Snapshot::short_help() const
+{
+    return QStringLiteral("Take a snapshot of an instance");
+}
+
+QString cmd::Snapshot::description() const
+{
+    return QStringLiteral("Take a snapshot of an instance that can later be restored to recover the current state.");
+}
+
+mp::ParseCode cmd::Snapshot::parse_args(mp::ArgParser* parser)
+{
+    parser->addPositionalArgument("instance", "The instance to take a snapshot of.");
+    QCommandLineOption name_opt({"n", "name"},
+                                "An optional name for the snapshot, subject to the same validity rules as instance "
+                                "names (see `help launch`). Default: \"snapshotN\", where N is one plus the "
+                                "number of snapshots that were ever taken for <instance>.",
+                                "name");
+    QCommandLineOption comment_opt{{"comment", "c", "m"},
+                                   "An optional free comment to associate with the snapshot.",
+                                   "comment"};
```

Comment:
I am not aware of a precedent with comments, but I can think of a few precedents for arguments that may need quoting due to characters that the shell would otherwise interpret. For instance, network names on Windows often have spaces, so the argument to `--network` needs quoting there. Path arguments in `mount` or `transfer` (can have spaces or other weird chars). Where settings receive free strings, they may need quoting.

I believe this in the shell's territory, in that it is knowledge about using the command line in general, not Multipass itself (Multipass never sees those quotes). That said, I do think we can go a little out of our way and add some tip to help users avoid the pitfall. But notice that such wrong usage will be detected anyway:

```
$ multipass snapshot temperate-violetear --comment This is a foo snapshot
Too many arguments supplied
```

Reply:
How about *An optional free comment to associate with the snapshot. (Tip: quote the text to avoid spaces being parsed by your shell)* ?

To answer your last question: yes, I think it is well known with CLI tools that comments like this should be enclosed in quotes. However, it doesn't hurt to make it easy for people who haven't learnt that yet :slightly_smiling_face:

Reply:
I like the tip idea and we should expand that elsewhere in the near future.  So yes, please add that here.  I like us being helpful as we've seen a number of users who are not comfortable at all with the command line.  The GUI should help alleviate some of that, but still.